devdoc/doublylinkedlist_requirements.md

```diff
@@ -20,13 +20,43 @@ typedef struct DLIST_ENTRY_TAG
     struct DLIST_ENTRY_TAG *Blink;
 } DLIST_ENTRY, *PDLIST_ENTRY;
 
+#define DLIST_FIND_RESULT_VALUES \
+    DLIST_FIND_RESULT_FOUND, \
+    DLIST_FIND_RESULT_NOT_FOUND, \
+    DLIST_FIND_RESULT_ERROR
+
+MU_DEFINE_ENUM(DLIST_FIND_RESULT, DLIST_FIND_RESULT_VALUES);
+
+#define DLIST_MATCH_FUNCTION_RESULT_VALUES \
+    DLIST_MATCH_FUNCTION_MATCHING, \
+    DLIST_MATCH_FUNCTION_NOT_MATCHING, \
+    DLIST_MATCH_FUNCTION_ERROR
+
+MU_DEFINE_ENUM(DLIST_MATCH_FUNCTION_RESULT, DLIST_MATCH_FUNCTION_RESULT_VALUES);
+
+#define DLIST_CONDITION_FUNCTION_RESULT_VALUES \
+    DLIST_CONDITION_FUNCTION_TRUE, \
+    DLIST_CONDITION_FUNCTION_FALSE, \
+    DLIST_CONDITION_FUNCTION_ERROR
+
+MU_DEFINE_ENUM(DLIST_CONDITION_FUNCTION_RESULT, DLIST_CONDITION_FUNCTION_RESULT_VALUES);
+
+typedef DLIST_MATCH_FUNCTION_RESULT (*DLIST_MATCH_FUNCTION)(PDLIST_ENTRY listEntry, const void* matchContext);
+typedef DLIST_CONDITION_FUNCTION_RESULT (*DLIST_CONDITION_FUNCTION)(PDLIST_ENTRY listEntry, const void* conditionContext, bool* continueProcessing);
+typedef int (*DLIST_ENTRY_DESTROY_FUNCTION)(PDLIST_ENTRY listEntry, const void* destroyContext);
+typedef int (*DLIST_ACTION_FUNCTION)(PDLIST_ENTRY listEntry, const void* actionContext, bool* continueProcessing);
+
 extern void DList_InitializeListHead(PDLIST_ENTRY listHead);
 extern int DList_IsListEmpty(const PDLIST_ENTRY listHead);
 extern void DList_InsertTailList(PDLIST_ENTRY listHead, PDLIST_ENTRY listEntry);
 extern void DList_InsertHeadList(PDLIST_ENTRY listHead, PDLIST_ENTRY entry);
 extern void DList_AppendTailList(PDLIST_ENTRY listHead, PDLIST_ENTRY ListToAppend);
 extern int DList_RemoveEntryList(PDLIST_ENTRY listEntry);
 extern PDLIST_ENTRY DList_RemoveHeadList(PDLIST_ENTRY listHead);
+extern DLIST_FIND_RESULT DList_Find(PDLIST_ENTRY listHead, DLIST_MATCH_FUNCTION matchFunction, const void* matchContext, PDLIST_ENTRY* foundEntry);
+extern int DList_RemoveIf(PDLIST_ENTRY listHead, DLIST_CONDITION_FUNCTION conditionFunction, const void* conditionContext, DLIST_ENTRY_DESTROY_FUNCTION destroyFunction, const void* destroyContext);
+extern int DList_ForEach(PDLIST_ENTRY listHead, DLIST_ACTION_FUNCTION actionFunction, const void* actionContext);
```

Comment:
>DList_ForEach [](http://example.com/codeflow?start=11&length=13)

I don't think ForEach should take an "actionFunction". Keeping it very easy, it would just enumerate the DLIST_ENTRYs in the list. It is up to the user to perform "some" action on it.

This all sounds very cpluspus-ish


Oh wait, maybe it is just poetry. I saw "aCTIOn" and I thought... "it does something". But maybe it just enumerates.

./consider ACTION => ENUMERATE.

Reply:
keeping it as action.